libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerAction.cs

```diff
@@ -0,0 +1,143 @@
+            this.KbId = kbId;
+            this.HostName = hostName;
+            this.EndpointKey = endpointKey;
+            this.httpClient = httpClient;
+        }
+
+        [JsonConstructor]
+        public QnAMakerAction([CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int sourceLineNumber = 0)
+            : base()
+        {
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+        }
+
+        [JsonProperty("kbId")]
+        public string KbId { get; set; }
+
+        [JsonProperty("hostname")]
+        public string HostName { get; set; }
+
+        [JsonProperty("endpointKey")]
+        public string EndpointKey { get; set; }
+
+        [JsonProperty("threshold")]
+        public float Threshold { get; set; }
+
+        [JsonProperty("noAnswer")]
+        public string NoAnswer { get; set; }
+
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var endpoint = new QnAMakerEndpoint
+            {
+                EndpointKey = this.EndpointKey,
+                Host = this.HostName,
+                KnowledgeBaseId = this.KbId
+            };
+
+            var qnamakerOptions = new QnAMakerOptions
+            {
+                ScoreThreshold = this.Threshold
+            };
+
+            if (qnamaker == null)
+            {
+                qnamaker = new QnAMaker(endpoint, qnamakerOptions, httpClient);
+            }
+
+            if (dc == null)
+            {
+                throw new ArgumentNullException(nameof(dc));
+            }
+
+            if (dc.Context?.Activity?.Type != ActivityTypes.Message)
+            {
+                return Dialog.EndOfTurn;
+            }
+
+            return await ExecuteAdaptiveQnAMakerDialog(dc, qnamaker, qnamakerOptions, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<DialogTurnResult> ExecuteQnAMakerDialog(DialogContext dc, QnAMaker qnaMaker, QnAMakerOptions qnamakerOptions, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var questionResults = await qnamaker.GetAnswersAsync(dc.Context, qnamakerOptions).ConfigureAwait(false);
+
+            // TODO: Get active learning flag: Need to update when this support is added to SDK
+            var isActiveLearningEnabled = true;
+
+            var queryResults = questionResults;
+
+            var noAnswerMsg = this.NoAnswer ?? "No QnAMaker answers found.";
+
+            if (queryResults == null || queryResults.Length == 0)
+            {
+                await dc.Context.SendActivityAsync(noAnswerMsg, cancellationToken: cancellationToken).ConfigureAwait(false);
+                return await dc.EndDialogAsync(false, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
```

Comment:
Some comments around this entire code would be incredibly helpful to understand why we are doing what we are doing,.

Reply:
I have added few comments.